File: battlebot/src/motor_controller.py
# motor_controller.py - Motor control for battlebot using ESC Motor class

import machine
import time
from config import *
from classes.esc_motor import Motor

class MotorController:
    """
    Controls drive motors and weapon using ESC Motor class.
    Includes safety features like failsafe and deadband.
    """

    def __init__(self):
        # Initialize ESC motors
        self.motor_left = Motor(
            signal_pin=PIN_MOTOR_LEFT_PWM,
            freq_hz=MOTOR_PWM_FREQ,
            min_us=MOTOR_MIN_US,
            mid_us=MOTOR_MID_US,
            max_us=MOTOR_MAX_US
        )

        self.motor_right = Motor(
            signal_pin=PIN_MOTOR_RIGHT_PWM,
            freq_hz=MOTOR_PWM_FREQ,
            min_us=MOTOR_MIN_US,
            mid_us=MOTOR_MID_US,
            max_us=MOTOR_MAX_US
        )

        self.weapon = Motor(
            signal_pin=PIN_WEAPON_PWM,
            freq_hz=MOTOR_PWM_FREQ,
            min_us=WEAPON_MIN_US,
            mid_us=WEAPON_MID_US,
            max_us=WEAPON_MAX_US
        )

        # Current motor states (stored as -100 to 100 for logging compatibility)
        self.left_speed = 0
        self.right_speed = 0
        self.weapon_speed = 0

        # Failsafe tracking
        self.last_command_time = time.ticks_ms()
        self.armed = False

        # Initialize stopped
        print("Arming ESCs...")
        self.motor_left.arm(hold_seconds=MOTOR_ARM_TIME)
        self.motor_right.arm(hold_seconds=0.1)  # Already waited for left
        self.weapon.arm(hold_seconds=0.1)
        print("ESCs armed and ready")

        self.stop_all()

    def set_left_motor(self, speed):
        """
        Set left motor speed.

        Args:
            speed: -100 to 100 (negative = reverse)
        """
        speed = self._apply_deadband(speed)
        speed = max(-MOTOR_MAX_SPEED, min(MOTOR_MAX_SPEED, speed))

        # Convert -100 to 100 range to 0.0 to 1.0 throttle
        # Negative values are reverse (if ESC supports bidirectional)
        # For forward-only ESCs, use absolute value
        if MOTOR_BIDIRECTIONAL:
            # Map -100..100 to 0.0..1.0 (0.5 = stopped, <0.5 = reverse, >0.5 = forward)
            throttle = (speed + 100) / 200.0
        else:
            # Forward only - use absolute value
            throttle = abs(speed) / 100.0

        self.motor_left.set_throttle(throttle)
        self.left_speed = speed
        self._update_command_time()

    def set_right_motor(self, speed):
        """
        Set right motor speed.

        Args:
            speed: -100 to 100 (negative = reverse)
        """
        speed = self._apply_deadband(speed)
        speed = max(-MOTOR_MAX_SPEED, min(MOTOR_MAX_SPEED, speed))

        if MOTOR_BIDIRECTIONAL:
            throttle = (speed + 100) / 200.0
        else:
            throttle = abs(speed) / 100.0

        self.motor_right.set_throttle(throttle)
        self.right_speed = speed
        self._update_command_time()

    def set_weapon(self, speed):
        """
        Set weapon motor speed.

        Args:
            speed: 0 to 100 (weapon typically one direction)
        """
        if not self.armed:
            speed = 0

        speed = max(0, min(MOTOR_MAX_SPEED, speed))

        # Weapon is typically forward only
        throttle = speed / 100.0
        self.weapon.set_throttle(throttle)
        self.weapon_speed = speed
        self._update_command_time()

    def set_tank_drive(self, left, right):
        """Convenience method for tank drive control"""
        self.set_left_motor(left)
        self.set_right_motor(right)

    def set_arcade_drive(self, throttle, turn):
        """
        Arcade drive mixing.

        Args:
            throttle: -100 to 100 (forward/back)
            turn: -100 to 100 (left/right)
        """
        left = throttle + turn
        right = throttle - turn

        # Normalize if over 100
        max_val = max(abs(left), abs(right))
        if max_val > 100:
            left = (left / max_val) * 100
            right = (right / max_val) * 100

        self.set_tank_drive(int(left), int(right))

    def arm(self):
        """Arm the weapon - must be called before weapon will spin"""
        self.armed = True
        print("Weapon ARMED")

    def disarm(self):
        """Disarm the weapon - weapon will not spin"""
        self.armed = False
        self.set_weapon(0)
        print("Weapon DISARMED")

    def stop_all(self):
        """Emergency stop - all motors to zero"""
        self.set_left_motor(0)
        self.set_right_motor(0)
        self.set_weapon(0)
        self.disarm()

    def check_failsafe(self):
        """
        Check if failsafe should trigger (no commands received).
        Returns True if failsafe activated.
        """
        if not ENABLE_FAILSAFE:
            return False

        if time.ticks_diff(time.ticks_ms(), self.last_command_time) > FAILSAFE_TIMEOUT_MS:
            self.stop_all()
            return True

        return False

    def _apply_deadband(self, speed):
        """Apply deadband to eliminate stick drift"""
        if abs(speed) < MOTOR_DEADBAND:
            return 0
        return speed

    def _update_command_time(self):
        """Update last command timestamp for failsafe"""
        self.last_command_time = time.ticks_ms()

    def get_status(self):
        """Get current motor status"""
        return {
            'left': self.left_speed,
            'right': self.right_speed,
            'weapon': self.weapon_speed,
            'armed': self.armed,
            'failsafe_ok': not self.check_failsafe()
        }

File: src/hardware/motor_controller.c
// =============================================================================
// Motor Controller Implementation
// Tank/arcade drive mixing, weapon control, and failsafe
// =============================================================================

#include "motor_controller.h"
#include <stdio.h>
#include <stdlib.h>
#include "pico/time.h"

// Helper: Apply deadband to eliminate stick drift
static int apply_deadband(int value) {
    if (abs(value) < MOTOR_DEADBAND) {
        return 0;
    }
    return value;
}

// Helper: Clamp value to range
static int clamp(int value, int min, int max) {
    if (value < min) return min;
    if (value > max) return max;
    return value;
}

// Helper: Update command timestamp for failsafe
static void update_command_time(motor_controller_t* mc) {
    mc->last_command_time_ms = to_ms_since_boot(get_absolute_time());
    mc->failsafe_triggered = false;
}

void motor_controller_init(motor_controller_t* mc) {
    printf("Initializing motor controller...\n");

    // Initialize drive motors (4WD tank drive)
    motor_init(&mc->motor_left_front, PIN_MOTOR_LEFT_FRONT, DRIVE_MIN_US, DRIVE_MID_US, DRIVE_MAX_US);
    motor_init(&mc->motor_left_back, PIN_MOTOR_LEFT_BACK, DRIVE_MIN_US, DRIVE_MID_US, DRIVE_MAX_US);
    motor_init(&mc->motor_right_front, PIN_MOTOR_RIGHT_FRONT, DRIVE_MIN_US, DRIVE_MID_US, DRIVE_MAX_US);
    motor_init(&mc->motor_right_back, PIN_MOTOR_RIGHT_BACK, DRIVE_MIN_US, DRIVE_MID_US, DRIVE_MAX_US);

    // Initialize weapon motor
    motor_init(&mc->weapon, PIN_WEAPON, WEAPON_MIN_US, WEAPON_MID_US, WEAPON_MAX_US);

    // Initialize state
    mc->left_speed = 0;
    mc->right_speed = 0;
    mc->weapon_speed = 0;
    mc->weapon_armed = false;
    mc->last_command_time_ms = to_ms_since_boot(get_absolute_time());
    mc->failsafe_triggered = false;

    // Stop drive motors but arm weapon (no safety arming required)
    motor_stop(&mc->motor_left_front, MOTOR_BIDIRECTIONAL);
    motor_stop(&mc->motor_left_back, MOTOR_BIDIRECTIONAL);
    motor_stop(&mc->motor_right_front, MOTOR_BIDIRECTIONAL);
    motor_stop(&mc->motor_right_back, MOTOR_BIDIRECTIONAL);
    mc->left_speed = 0;
    mc->right_speed = 0;

    // Auto-arm weapon
    motor_controller_arm_weapon(mc);

    printf("Motor controller ready (4WD tank drive, weapon ARMED)\n");
}

void motor_controller_set_left(motor_controller_t* mc, int speed) {
    speed = apply_deadband(speed);
    speed = clamp(speed, -MOTOR_MAX_SPEED, MOTOR_MAX_SPEED);

    // Drive both left motors together
    motor_set_speed(&mc->motor_left_front, speed, MOTOR_BIDIRECTIONAL);
    motor_set_speed(&mc->motor_left_back, speed, MOTOR_BIDIRECTIONAL);
    mc->left_speed = speed;
    update_command_time(mc);
}

void motor_controller_set_right(motor_controller_t* mc, int speed) {
    speed = apply_deadband(speed);
    speed = clamp(speed, -MOTOR_MAX_SPEED, MOTOR_MAX_SPEED);

    // Drive both right motors together
    motor_set_speed(&mc->motor_right_front, speed, MOTOR_BIDIRECTIONAL);
    motor_set_speed(&mc->motor_right_back, speed, MOTOR_BIDIRECTIONAL);
    mc->right_speed = speed;
    update_command_time(mc);
}

void motor_controller_tank_drive(motor_controller_t* mc, int left, int right) {
    motor_controller_set_left(mc, left);
    motor_controller_set_right(mc, right);
}

void motor_controller_arcade_drive(motor_controller_t* mc, int throttle, int turn) {
    // Apply deadband to inputs
    throttle = apply_deadband(throttle);
    turn = apply_deadband(turn);

    // Arcade mixing: throttle controls forward/back, turn controls rotation
    int left = throttle + turn;
    int right = throttle - turn;

    // Normalize if over max speed (preserve ratio)
    int max_val = abs(left);
    if (abs(right) > max_val) {
        max_val = abs(right);
    }

    if (max_val > MOTOR_MAX_SPEED) {
        left = (left * MOTOR_MAX_SPEED) / max_val;
        right = (right * MOTOR_MAX_SPEED) / max_val;
    }

    motor_controller_set_left(mc, left);
    motor_controller_set_right(mc, right);
}

void motor_controller_set_weapon(motor_controller_t* mc, int speed) {
    speed = clamp(speed, 0, MOTOR_MAX_SPEED);

    // Weapon uses same ESC type as drive motors
    motor_set_speed(&mc->weapon, speed, MOTOR_BIDIRECTIONAL);
    mc->weapon_speed = speed;
    update_command_time(mc);
}

void motor_controller_arm_weapon(motor_controller_t* mc) {
    mc->weapon_armed = true;
    motor_arm(&mc->weapon);
    printf("*** WEAPON ARMED ***\n");
}

void motor_controller_disarm_weapon(motor_controller_t* mc) {
    bool was_armed = mc->weapon_armed;
    mc->weapon_armed = false;
    motor_disarm(&mc->weapon);

    // Stop weapon immediately
    motor_set_speed(&mc->weapon, 0, MOTOR_BIDIRECTIONAL);
    mc->weapon_speed = 0;

    if (was_armed) {
        printf("*** WEAPON DISARMED ***\n");
    }
}

bool motor_controller_is_weapon_armed(motor_controller_t* mc) {
    return mc->weapon_armed;
}

void motor_controller_stop_all(motor_controller_t* mc) {
    // Stop all 4 drive motors
    motor_stop(&mc->motor_left_front, MOTOR_BIDIRECTIONAL);
    motor_stop(&mc->motor_left_back, MOTOR_BIDIRECTIONAL);
    motor_stop(&mc->motor_right_front, MOTOR_BIDIRECTIONAL);
    motor_stop(&mc->motor_right_back, MOTOR_BIDIRECTIONAL);
    mc->left_speed = 0;
    mc->right_speed = 0;

    // Disarm and stop weapon
    motor_controller_disarm_weapon(mc);
}

bool motor_controller_check_failsafe(motor_controller_t* mc) {
    if (!FAILSAFE_ENABLED) {
        return false;
    }

    uint32_t now = to_ms_since_boot(get_absolute_time());
    uint32_t elapsed = now - mc->last_command_time_ms;

    if (elapsed > FAILSAFE_TIMEOUT_MS) {
        if (!mc->failsafe_triggered) {
            // Only print if motors were actually running
            if (mc->left_speed != 0 || mc->right_speed != 0 || mc->weapon_speed != 0) {
                printf("Failsafe: motors stopped\n");
            }
            motor_controller_stop_all(mc);
            mc->failsafe_triggered = true;
        }
        return true;
    }

    return false;
}

void motor_controller_get_status(motor_controller_t* mc, int* left, int* right, int* weapon, bool* armed) {
    if (left) *left = mc->left_speed;
    if (right) *right = mc->right_speed;
    if (weapon) *weapon = mc->weapon_speed;
    if (armed) *armed = mc->weapon_armed;
}
